File: coding-interviewes/coding_interview_150/java/HIndex.java

/* H-Index — sort ascending & scan. O(n log n) */
import java.util.Arrays;
public class HIndex {
    public static int hIndex(int[] c){
        Arrays.sort(c);
        int n=c.length;
        for (int i=0;i<n;i++){ int papers=n-i; if (c[i]>=papers) return papers; }
        return 0;
    }
    public static void main(String[] args){
        System.out.println(hIndex(new int[]{3,0,6,1,5}));
    }
}

File: coding-interviewes/coding_interview_150/scala/HIndex.scala

/* H-Index — sort & scan. O(n log n) */
object HIndex {
  def hIndex(c:Array[Int]):Int = {
    scala.util.Sorting.quickSort(c)
    val n=c.length
    for (i <- 0 until n){
      val papers = n - i
      if (c(i) >= papers) return papers
    }
    0
  }
  def main(args:Array[String]):Unit = {
    println(hIndex(Array(3,0,6,1,5)))
  }
}
